csharp/ql/src/semmle/code/csharp/dataflow/FrameworkDataFlow.qll

```diff
@@ -0,0 +1,502 @@
+module FlowSource {
+  FlowSource remote() { result = TRemoteFlowSource() }
+
+  FlowSource local() { result = TLocalFlowSource() }
+
+  FlowSource stored() { result = TStoredFlowSource() }
+}
+
+/** A flow-sink specification. */
+class FlowSink extends TFlowSink {
+  final string toString() {
+    this = TRemoteFlowSink() and
+    result = "remote"
+    or
+    this = THtmlFlowSink() and
+    result = "html"
+    or
+    this = TEmailFlowSink() and
+    result = "email"
+    or
+    this = TSqlFlowSink() and
+    result = "sql"
+    or
+    this = TCommandFlowSink() and
+    result = "command"
+  }
+}
+
+/** Provides predicates for constructing flow-sink specifications. */
+module FlowSink {
+  FlowSink remote() { result = TRemoteFlowSink() }
+
+  FlowSink html() { result = THtmlFlowSink() }
+
+  FlowSink email() { result = TEmailFlowSink() }
+
+  FlowSink sql() { result = TSqlFlowSink() }
+
+  FlowSink command() { result = TCommandFlowSink() }
+}
+
+/** A data-flow model for a given framework. */
+abstract class FrameworkDataFlow extends string {
+  bindingset[this]
+  FrameworkDataFlow() { any() }
+
+  /**
+   * Holds if data may flow from `input` to `output` when calling `c`.
+   *
+   * `inputStack` describes the contents that is popped from the access
+   * path from the input and `outputStack` describes the contents that
+   * is pushed onto the resulting access path.
+   *
+   * `preservesValue` indicates whether this is a value-preserving step
+   * or a taint-step.
+   */
+  pragma[nomagic]
+  predicate hasSummary(
+    SourceDeclarationCallable c, SummaryInput input, ContentStack inputStack, SummaryOutput output,
+    ContentStack outputStack, boolean preservesValue
+  ) {
+    none()
+  }
+
+  /**
+   * Holds if the content stack obtained by pushing `head` onto `tail` is
+   * needed for a summary specified by `hasSummary()`.
+   *
+   * This predicate is needed for QL technical reasons only (the IPA type used
+   * to represent content stacks needs to be bounded).
+   */
+  pragma[nomagic]
+  predicate requiresContentStack(Content head, ContentStack tail) { none() }
+
+  /**
+   * Holds if values stored inside `content` are cleared on objects passed as
+   * arguments of type `input` to calls that target `c`.
+   */
+  pragma[nomagic]
+  predicate clearsContent(SourceDeclarationCallable c, SummaryInput input, Content content) {
+    none()
+  }
+
+  /** Holds if `n` is a flow source of type `source. */
+  pragma[nomagic]
+  predicate hasSource(DataFlow::Node n, FlowSource source) { none() }
+
+  /** Holds if `n` is a flow sink of type `sink. */
+  pragma[nomagic]
+  predicate hasSink(DataFlow::Node n, FlowSink sink) { none() }
+}
+
+module Examples {
+  private import semmle.code.csharp.frameworks.system.Text
+
+  /** Data flow for `System.Text.StringBuilder`. */
+  class StringBuilderDataFlow extends FrameworkDataFlow {
+    StringBuilderDataFlow() { this = "System.Text.StringBuilder" }
```

Comment:
Isn't this the answer? (From the PR description)

> The reason for defining `FrameworkDataFlow` as an abstract class that extends `string`, as opposed to a unit type, is that we can use it to provide some numbers on how much coverage we have for a given framework, e.g.
> 
> ```ql
> int countCallables(FrameworkDataFlow framework) {
>   result =
>     count(Callable c | framework.hasSummary(c, _, _, _, _, _) or framework.clearsContent(c, _, _))
> }
> ```

Reply:
I think it can be useful, for example if we want to have numbers on how much coverage we have for a given framework (see PR description for an example). Once the summaries have been compiled down to atomic steps we will no longer carry around the string column.

Reply:
I wouldn't call stringbuilder a framework, so methodology-wise it was a bad example, I merely included it to show examples of flow summaries.

Reply:
I'm not convinced that the framework string is going to be useful. A class or method can only be declared as belonging to a framework when it carries data flow. There's no way to state that a class is part of some framework but has no data flow. If that `count(...)` example tells us that a certain snapshot has 20 classes from the ASP.NET framework modelled, is that a lot or a little? It would depend on the number of classes from ASP.NET that are _not_ modelled.

If we can write one line of QL centrally to declare that all classes under the `System.Web` hierarchy belong to the ASP.NET framework, then we'll have described both the classes that have data-flow models and the classes that don't. That seems both more general and more brief than repeating the ASP.NET name in every data-flow model of a `System.Web` class.

We used to have the Tech Inventory feature to figure out which elements come from which frameworks. I'm sure that feature can still be brought to life if we want.

Reply:
I agree that the string may not be relevant for flow summaries, but for sources and sinks I don't think we can necessarily map them easily back to their framework.